### DISCUSSION TITLE
Vim mode in Codemirror 6

### DISCUSSION THREAD

**Discussion by kyaso** [Zettlr/Zettlr#3751] [General]
I was wondering, whether CM6 come with the Vim extension by default?

If not, we could use this plugin: https://github.com/replit/codemirror-vim

The plugin was also mentioned by the CM dev: https://github.com/codemirror/dev/issues/79#issuecomment-1149879825

**Comment by nathanlesage** [comment]
I've seen that and I know that we'll have to do a ton of work to enable the different modes again, but the keybindings work much better now.

Would you mind taking this task on your plate? This way I'd first finish the CM6 migration PR and once that runs smoothly, we can add the three keymaps for default/emacs/vim. I guess only the vim-keymap should be somewhat more difficult, right?

I think it should even be possible to start with the VIM keymap, as it needs to be an extension anyways that I can pull in to documents.

Thoughts?

**Reply by kyaso** [reply]
Sure, I can give it a shot once the CM6 migration is ready.

Though I can't fully judge the required effort. Could you maybe elaborate on what you mean by "we'll have to do a ton of work to enable the different modes again"? Looking at that Vim extension above, doesn't it look relatively straightforward to integrate, or am I missing something?

**Reply by nathanlesage** [reply]
Well, the Emacs and default modes are basically just a set of keybindings that trigger a few commands on the editor. The vim mode, however, has a much bigger internal state (since you can do, e.g., "yanking" or other fancy things that require a state). That being said, the key bindings have to be implemented so that depending on the state certain actions are taken. I think I saw an implementation on how that worked on the link you provided. But I mean, as long as the actual functionality of vim is inside the plugin, then the efforts to actually implement that shouldn't be larger than writing a few functions (which is still more complicated as the Emacs mode, for sure).

**Comment by tobarja** [comment]
Did this go anywhere? I see that I can enable vim mode, and it mostly works.

Is there a list of key bindings in code somewhere? I can write up the beginnings of a doc page.

**Reply by nathanlesage** [reply]
The vim key map is here in this file: https://github.com/replit/codemirror-vim/blob/master/src/vim.js
